query-engine/js-connectors/src/proxy.rs

```diff
@@ -289,16 +294,143 @@ fn js_planetscale_value_to_quaint(json_value: serde_json::Value, column_type: Co
     }
 }
 
-impl From<JSResultSet> for QuaintResultSet {
-    fn from(mut js_result_set: JSResultSet) -> Self {
+fn js_neon_value_to_quaint(json_value: serde_json::Value, column_type: ColumnType) -> QuaintValue<'static> {
+    //  Note for the future: it may be worth revisiting how much bloat so many panics with different static
+    // strings add to the compiled artefact, and in case we should come up with a restricted set of panic
+    // messages, or even find a way of removing them altogether.
+    match column_type {
+        ColumnType::Int32 => match json_value {
+            serde_json::Value::Number(n) => {
+                // n.as_i32() is not implemented, so we need to downcast from i64 instead
+                QuaintValue::int32(n.as_i64().expect("number must be an i32") as i32)
+            }
+            serde_json::Value::Null => QuaintValue::Int32(None),
+            mismatch => panic!("Expected an i32 number, found {:?}", mismatch),
+        },
+        ColumnType::Int64 => match json_value {
+            serde_json::Value::String(s) => {
+                let n = s.parse::<i64>().expect("string-encoded number must be an i64");
+                QuaintValue::int64(n)
+            }
+            serde_json::Value::Null => QuaintValue::Int64(None),
+            mismatch => panic!("Expected a string, found {:?}", mismatch),
+        },
+        ColumnType::Float => match json_value {
+            // n.as_f32() is not implemented, so we need to downcast from f64 instead
+            serde_json::Value::Number(n) => QuaintValue::float(n.as_f64().expect("number must be a f32") as f32),
```

Comment:
Unsafe cast, some overflowing can happen here

Reply:
How would you recommend tackling this? In TypeScript, we tag MySQL's type `FLOAT_32` (as returned by PlanetScale's query result) with `ColumnType::Float`. So, the underlying precondition is that `serde_json::Value::Number(n)` is compatible with Rust's `f32`.

In Rust's MySQL driver (see [here](https://github.com/prisma/prisma-engines/blob/032ea5987050b4ef0b92272bc00f1e50ff02798a/quaint/src/connector/mysql/conversion.rs#L277)), we have a better type signature in the pattern matching, as we actually get handed a `f32` directly, but conceptually there's a similar underlying idea: the database gives us a value and tells us which type it has, and we trust the database.

Reply:
I understand there's some trust going on at the protocol level and I understand there's not much we can do in case it ever overflows. However, having an error would at least tell us what when wrong if it ever does.

If the f64 overflows when cast to an f32, [its value should be set to Infinity](https://stackoverflow.com/questions/72247741/how-to-convert-a-f64-to-a-f32). Therefore, adding the following checks should already catch some potential issues:

```rust
fn f64_to_f32(x: f64) -> f32 {
    let y = x as f32;

    assert_eq!(
        x.is_finite(),
        y.is_finite(),
        "f32 overflow during conversion"
    );

    y
}
```

Reply:
> there's some trust going on at the protocol level

Indeed.

> adding the following checks should already catch some potential issues

That sounds like a very reasonable addition. Thank you!